sklearn/tree/_tree.pyx

```diff
@@ -1132,3 +1132,296 @@ cdef class Tree:
+    original tree are copied into the pruned tree.
+
+    Parameters
+    ----------
+    tree : Tree
+        Location to place the pruned tree
+    orig_tree : Tree
+        Original tree
+    ccp_alpha : positive double
+        Complexity parameter. The subtree with the largest cost complexity
+        that is smaller than ``ccp_alpha`` will be choosen. By default,
+        no pruning is performed.no pruning is performed.
+    """
+
+    cdef:
+        SIZE_t i
+        SIZE_t n_nodes = orig_tree.node_count
+        # prior probability using weighted samples
+        DOUBLE_t[:] weighted_n_node_samples = orig_tree.weighted_n_node_samples
+        DOUBLE_t total_sum_weights = weighted_n_node_samples[0]
+        DOUBLE_t[:] impurity = orig_tree.impurity
+        # weighted impurity of each node
+        DOUBLE_t[:] r_node = np.empty(shape=n_nodes, dtype=np.float64)
+
+        SIZE_t[:] child_l = orig_tree.children_left
+        SIZE_t[:] child_r = orig_tree.children_right
+        SIZE_t[:] parent = np.zeros(shape=n_nodes, dtype=np.intp)
+        unsigned char[:] leaves_in_subtree = np.zeros(
+            shape=n_nodes, dtype=np.uint8)
+
+        # Only uses the start and parent variables
+        Stack stack = Stack(INITIAL_STACK_SIZE)
+        StackRecord stack_record
+        int rc = 0
+        SIZE_t node_idx
+
+        SIZE_t[:] n_leaves = np.zeros(shape=n_nodes, dtype=np.intp)
+        DOUBLE_t[:] r_branch = np.zeros(shape=n_nodes, dtype=np.float64)
+        DOUBLE_t current_r
+        SIZE_t leaf_idx
+        SIZE_t parent_idx
+
+        # candidate nodes that can be pruned
+        bint[:] candidate_nodes = view.array(
+            shape=(n_nodes, ), itemsize=sizeof(bint), format="i")
+        # nodes in subtree
+        bint[:] in_subtree = view.array(
+            shape=(n_nodes, ), itemsize=sizeof(bint), format="i")
+        DOUBLE_t[:] g_node = np.zeros(shape=n_nodes, dtype=np.float64)
+        SIZE_t pruned_branch_node_idx
+        DOUBLE_t subtree_alpha
+        DOUBLE_t current_ccp_alpha
+        SIZE_t child_l_idx
+        SIZE_t child_r_idx
+        SIZE_t n_pruned_leaves
+        DOUBLE_t r_diff
+        DOUBLE_t max_float64 = np.finfo(np.float64).max
+        SIZE_t capacity = 0
+
+    # find parent node ids and leaves
+    with nogil:
+
+        for i in range(r_node.shape[0]):
+            r_node[i] = (
+                weighted_n_node_samples[i] * impurity[i] / total_sum_weights)
+
+        for i in range(n_nodes):
+            leaves_in_subtree[i] = 0
+
+        # Push root node, using StackRecord.start as node id
+        rc = stack.push(0, 0, 0, -1, 0, 0, 0)
+        if rc == -1:
+            with gil:
+                raise MemoryError()
+
+        while not stack.is_empty():
+            stack.pop(&stack_record)
+            node_idx = stack_record.start
+            parent[node_idx] = stack_record.parent
+            if child_l[node_idx] == child_r[node_idx]:
```

Comment:
I guess comparing them to `TREE_LEAF` would be more explicit and easier to read.